packages/langium/src/parser/langium-parser.ts

```diff
@@ -247,12 +248,19 @@ export class LangiumParser {
             this.stack.pop();
         }
         if (obj.$type === DatatypeSymbol) {
-            const node = obj.$cstNode;
-            return node.text;
+            return this.constructDataTypeResult(obj.$cstNode);
         }
         return obj;
     }
 
+    private constructDataTypeResult(node: CstNode): ValueType {
+        let text = '';
+        for (const subNode of streamCst(node).filter(e => isLeafCstNode(e) && !e.hidden)) {
+            text += this.converter.convert(subNode.text, subNode).toString();
+        }
+        return this.converter.convert(text, node);
```

Comment:
Two more remarks on this part 😅:
 * Using `streamCst` does not consider intermediate calls to data type rules (a data type rule can use another data type rule).
 * Keyword tokens should not be passed to the value converter.

Proposal: transform to a recursive function to account for intermediate data type rules.